### PR TITLE
Take 2: Add, Edit, Delete settings for web apps and deployment slots (Explorer)

### PR DESCRIPTION
**Pull request by whiskeyjay**
A complete different design of #30.

image

image

image

### CODE REVIEW

**Comment by whiskeyjay** [comment]
@nturinski this PR contains the refactoring to the nodes so now you can easily get a node's parent node as well as the root tree data provider.

File: src/explorer/appSettingsNodes.ts [hunk 1]
```diff
@@ -0,0 +1,206 @@
...
+            contextValue: "applicationSettings",
+            iconPath: { 
+                light: path.join(__filename, '..', '..', '..', '..', 'resources', 'light', 'Settings_16x_vscode.svg'),
+                dark: path.join(__filename, '..', '..', '..', '..', 'resources', 'dark', 'Settings_16x_vscode.svg')
+            }
+        }
+    }
+
+    async getChildren(): Promise<NodeBase[]> {
+        this._settings = await this.getSettings();
```
Comment (EricJizbaMSFT) [inline src/explorer/appSettingsNodes.ts]: nit: both `this.getSettings()` and `this.generateTreeItems` are only used by this function. I would prefer to have the code directly in getChildren()

File: src/explorer/nodeBase.ts [hunk 2]
```diff
@@ -20,11 +19,17 @@ export class NodeBase {
         };
     }
 
-    async getChildren(azureAccount: AzureAccountWrapper): Promise<NodeBase[]> {
-        return [];
+    getTreeDataProvider<T extends TreeDataProvider<NodeBase>>(): T {
```
Comment (EricJizbaMSFT) [inline src/explorer/nodeBase.ts]: I think we can get rid of 'T' here. It will always be the same type in this codebase

**Comment by nturinski** [comment]
@whiskeyjay Suh-weeeet

File: src/explorer/appSettingsNodes.ts [hunk 3]
```diff
@@ -0,0 +1,206 @@
...
+        }
+    }
+
+    async edit(): Promise<void> {
+        const oldKey = this.key;
+        const newKey = await vscode.window.showInputBox({
+            ignoreFocusOut: true,
+            prompt: 'Enter setting key',
+            value: this.key,
+            validateInput: v => (<AppSettingsNode>this.getParentNode()).validateNewKeyInput(v, oldKey)
```
Comment (EricJizbaMSFT) [inline src/explorer/appSettingsNodes.ts]: I feel like all of these should either be `this.getParentNode<AppSettingsNode>()` or `<AppSettingsNode>this.parentNode`, right?

**Pull request merged**